app/web_framework/cluster_base_handler.py

```diff
@@ -62,6 +69,24 @@ def _handle_request_exception(self, ex):
         self.set_status(status_code)
         self.finish()
 
+    # For methods that annotate with `Tuple`, use `disable=invalid-sequence-index`
+    # This bug is fixed in https://github.com/PyCQA/astroid/commit/563031aaf13a44adc8db4f8d0ab8020d550aae00
+    # More information on the issue in https://github.com/PyCQA/pylint/issues/1212
+
+    def get_pagination_params(self) -> Tuple[int, int]:  # pylint: disable=invalid-sequence-index
+        offset = self.get_query_argument('offset', pagination_constants['DEFAULT_OFFSET'], True)
+        limit = self.get_query_argument('limit', pagination_constants['DEFAULT_LIMIT'], True)
+        return self._validate_arguments(int(offset), int(limit))
+
+    @staticmethod
+    def _validate_arguments(offset: int, limit: int) -> Tuple[int, int]:  # pylint: disable=invalid-sequence-index
+        offset = offset or pagination_constants['DEFAULT_OFFSET']
+        limit = limit or pagination_constants['DEFAULT_LIMIT']
+        # Negative offsets or limits are not allowed
+        non_negative_offset = max(offset, 0)
```

Comment:
Just to repeat my above comment, let's only support negative values for `offset` (not `limit`). And I think it's fine to save that for a later change. It seems like most of the conversation/comments/questions on this PR so far have been around the negative offset support, so that's a good sign that it would be good to split out -- get everything merged that everyone agrees on and save the highly-discussed piece for a separate change. This somewhat falls under Evan's point (b) above about splitting things out that are taking too long to implement, but I'd extend that to "too long to *complete* due to implementation or review" including a best effort to try to split out things that you expect to be more contentious or debatable. (And btw, I'll take any disagreement with this comment as proof of my point. 😝)

Also, with regards to your last question about wrap-around -- I don't think it makes sense to support wrap-around. In your example of `offset=-5&limit=20` I think it makes the most sense to only return the last five items. I think that's intuitive.

A clearer example of why wrap-around is not good is if we only had 2 builds total and asked for `offset=0&limit=8` would we return a list of 8 builds which are just the only two builds repeated four times (1, 2, 1, 2, 1, 2, 1, 2)? Bonkers!